synapse/storage/state.py

```diff
@@ -497,14 +540,20 @@ def _get_some_state_from_cache(self, group, types):
         def include(typ, state_key):
             valid_state_keys = type_to_key.get(typ, sentinel)
             if valid_state_keys is sentinel:
-                return False
+                return filtered_types is not None and typ not in filtered_types
             if valid_state_keys is None:
                 return True
             if state_key in valid_state_keys:
                 return True
             return False
 
-        got_all = is_all or not missing_types
```

Comment:
can't we just write `is_all or (not missing_types and filtered_types is not None)` rather than special-casing over a particular bug elsewhere in the algorithm?

Reply:
i find `is_all or (not missing_types and filtered_types is not None)` incredibly cryptic to reason about, to the extent i'm failing to convince myself it's even right.

For instance, if this is called with `types=[]` and `filtered_types=[whatever]`, then is_all could well be false (if the cache is empty), and missing_types would be falsey, and filtered_types would not be None... so got_all would be true, which is the wrong answer.

Which is why I very deliberately spelt out the special case we're handling here where `types=[]`, so we can't trust `missing_types`, which feels a lot easier to reason about.

Reply:
ok, but the fact that you are special-casing the empty list makes *me* suspect that there are bugs elsewhere.

and sorry, it should have been `is_all or (not missing_types and filtered_types is None)`. Maybe it should be:

```
got_all = is_all
if not got_all:
    # the cache is incomplete. We may still have got all the results we need, if 
    # we don't have any wildcards in the match list.
    if not missing_types and filtered_types is None:
        got_all = True
```